src/java/org/apache/cassandra/index/sai/disk/v1/Segment.java

```diff
@@ -230,4 +231,27 @@ public int proportionalAnnLimit(int limit, long totalRows)
         assert proportionalLimit >= 1 : proportionalLimit;
         return proportionalLimit;
     }
+
+    public long estimateMatchingRowsCount(Expression predicate, AbstractBounds<PartitionPosition> keyRange)
+    {
+        switch (predicate.getOp())
+        {
+            case MATCH:
+            case EQ:
+            case CONTAINS_KEY:
+            case CONTAINS_VALUE:
+                return metadata.estimateNumRowsMatchingExact(predicate.lower.value.raw);
+            case NOT_EQ:
+            case NOT_CONTAINS_KEY:
+            case NOT_CONTAINS_VALUE:
+                return metadata.numRows - metadata.estimateNumRowsMatchingExact(predicate.lower.value.raw);
+            case RANGE:
+                ByteBuffer lower = predicate.lower != null ? predicate.lower.value.raw : null;
+                ByteBuffer upper = predicate.upper != null ? predicate.upper.value.raw : null;
```

Comment:
I see we're using `raw` here instead of `encoded`, which has consequences for some data types (as you know). Are you able to simplify the logic because of the work you did with `encodeIfNeeded` method?

Reply:
If everything is raw, then I can compare `ByteBuffers` using `AbstractType` which is a natural comparator for raw serializations. In some cases I also need to deserialize the value to treat it as a number (needed for interpolation).

Reply:
I think this won't work correctly for `Integer` type though. When we index `0`, we end up adding this byte buffer to the `TermsDistribution`:

[-128, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

However, 0 is represented by this byte comparable:

`[-128, 0]`

I generated these arrays by looking in the `TermsDistribution` method and then also running the following where `value` is a 0 varint.

```
        var bc = IntegerType.instance.asComparableBytes(value, ByteComparable.Version.OSS41);
        var source = ByteSourceInverse.readBytes(bc);
        System.out.println(Arrays.toString(source));
```

Reply:
After looking closer, there are a few oddities.

1. `TypeUtil.encode` maps inet, decimal, and varint to a different encoding before we call the `Version.latest().onDiskFormat().encodeForTrie`.
2. `Version.latest().onDiskFormat().encodeForTrie` ultimately calls `TypeUtil.asComparableBytes` which skips encoding inet, decimal, and varint types.
3.  We also use `TypeUtil.toComparableBytes` when encoding in the `KDTreeSegmentBuilder`, which again skips encoding inet, decimal, and varint types.

I think we could remove the `asComparableBytes` and then the subsequent `fromComparableBytes` calls by adding a new `TypeUtil.fromComparableBytes` method that skips inet, decimal, and varint. This doesn't necessarily make things easier, but at least it is symmetric. Further, it'll provide for correct estimates on inet types, which actually produce different orderings (and therefore different results for range queries) whether based on the SAI index or the comparable byte representation.

We would still need to use the `encoded` value in this `estimateMatchingRowsCount` method though.

Reply:
My last comment for the day.

> We would still need to use the `encoded` value in this `estimateMatchingRowsCount` method though.

I think that using encoded would probably give us issues with the accuracy of `toBigDecimal` on varint, but not on decimal, since the one is a custom format while the other is a truncated decimal. Otherwise, inet could have issues in the relative distance between ipv4 and ipv6, but that is an issue for another day.

My biggest concerns really focus on what we're writing to disk since it takes longer to deliver improvements in those situations.

Reply:
After looking at this more today, I did not find a working solution. The primary issue for varint and decimal types are that they are encoded using unsigned bytes, but of course bytes are signed in java, so we end up with these byte arrays, which are not actually ordered in the right way, so equality checks will pass, but range checks won't.

In these outputs, I take a decimal or a varint, apply the TypeUtil.encode logic to it, and then get the array that results from `ByteSourceInverse.readBytes(byteComparable.asComparableBytes(version))`.

```
decimal encoding:
-1  -> [63, -1, 127, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
 0  -> [-128, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
 1  -> [-63, 1, -127, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

varint encoding:
-1 -> [127, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]
 0 ->  [-128, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
 1  ->  [-128, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]
```

I will try to spend a bit more time on this tomorrow.